Add a Unified Resources card component

bl-nero opened the pull request:
This partially addresses the "unified view" part of #27462.

avatus committed:
Update resourceKind method to return slice

avatus committed:
Adds a UnifiedResource watcher to the auth server. This watcher
will watch all the types that are displayed in the web UI and store
them in-memory to allow us to search/filter/query and get multiple
kinds returned at the same time.
[RFD](https://github.com/gravitational/teleport/pull/28162)

avatus committed:
Add UnifiedResourceComponent

avatus committed:
Add unifiedResourceWatcher to auth server

avatus committed:
Init unified resource watcher

avatus committed:
Convert to strings for error output

avatus committed:
Add license

bl-nero committed:
Merge branch 'master' into michaelmyers/unified_resource_watcher

bl-nero committed:
Add some tests

bl-nero committed:
Merge remote-tracking branch 'origin/master' into michaelmyers/unified_resource_watcher

bl-nero committed:
Fix a race condition and properly close the watcher

avatus committed:
add defer

avatus committed:
Add kubes to watcher

avatus committed:
Add uninit check

avatus committed:
Notify if stale

avatus committed:
Add SAML to watcher

avatus committed:
Update test and fix lint

avatus committed:
Merge branch 'master' into michaelmyers/unified_resource_watcher

avatus committed:
Fix license

avatus committed:
Add godocs

avatus committed:
Add Kind to ui structs

avatus committed:
Add Kind to ui structs

avatus committed:
Add FakePaginateParams

avatus committed:
Add FakePaginateParams

avatus committed:
Update ListUnifiedResources proto messages

avatus committed:
Update ListUnifiedResources proto messages

avatus committed:
Update uses of FakePaginate

avatus committed:
Update uses of FakePaginate

avatus committed:
Include type in seenmap for MatchResourceByFilters

avatus committed:
Include type in seenmap for MatchResourceByFilters

avatus committed:
Update TestAuthServer with unified resource watcher

avatus committed:
Update TestAuthServer with unified resource watcher

avatus committed:
CheckAndSetDefaults for ListUnifiedResourcesRequest

avatus committed:
CheckAndSetDefaults for ListUnifiedResourcesRequest

avatus committed:
Add ListUnifiedResources rpc

avatus committed:
Add ListUnifiedResources rpc

avatus committed:
Update protos

avatus committed:
Update protos

avatus committed:
Add test

avatus committed:
Add test

avatus committed:
Add docs and update tests

avatus committed:
Add docs and update tests

avatus committed:
Add more docs

avatus committed:
Add more docs

avatus committed:
Merge branch 'master' into michaelmyers/unified_resource_watcher

avatus committed:
Combine app types in switch

avatus committed:
Combine app types in switch

avatus committed:
Fix tests

avatus committed:
Fix tests

avatus committed:
Add kube access checker

avatus committed:
Add kube access checker

avatus committed:
Add remaining kinds

avatus committed:
Add remaining kinds

bl-nero committed:
Add a Unified Resource card component

bl-nero committed:
Lint fixes

zmb3 commented on web/packages/design/src/theme/sharedStyles.ts:
```diff
@@ -21,4 +21,9 @@ export const sharedStyles = {
     '0px 3px 5px -1px rgba(0, 0, 0, 0.2), 0px 6px 10px rgba(0, 0, 0, 0.14), 0px 1px 18px rgba(0, 0, 0, 0.12)',
   ],
+  breakpoints: {
+    mobile: 400 + 256, // sidebar width
```
Why not make a constant for sidebar width so we can drop the comments?

zmb3 commented on web/packages/teleport/src/Resources/ResourceCard.tsx:
```diff
@@ -0,0 +1,164 @@
+export const ResourceCard = ({ resource }: Props) => {
+  const name = agentName(resource);
+  const resIcon = agentIconName(resource);
+  const ResTypeIcon = agentTypeIcon(resource.kind);
```
```suggestion
  const resTypeIcon = agentTypeIcon(resource.kind);
```

Any particular reason this one is capitalized?

bl-nero replied:
Yes. In React, every component that will be rendered in JSX, needs to be capitalized; otherwise, it will be rendered as a plain HTML or SVG tag. I guess this is for compatibility with future HTML standards, as well as Web Components.

zmb3 commented on web/packages/teleport/src/Resources/ResourceCard.tsx:
```diff
@@ -0,0 +1,164 @@
+  );
+};
+
+function agentName(agent: AgentKind) {
```
Should these be named `resourceName`, `resourceDescription`, etc?

bl-nero replied:
I don't know. Maybe? We already use dual names for this, "agent" and "resource". But yeah, I can change it.

zmb3 commented on web/packages/teleport/src/Resources/ResourceCard.tsx:
```diff
@@ -0,0 +1,164 @@
+    case 'app':
+      return 'Application';
+    case 'db':
+      // agent.
```
```suggestion
```

bl-nero replied:
Oops :)

avatus committed:
Review feedback

avatus committed:
Add configurable queuesize to NewWatcher

kimlisa commented on web/packages/teleport/src/Resources/ResourceCard.tsx:
```diff
@@ -0,0 +1,164 @@
+  resource: AgentKind;
+};
+export const ResourceCard = ({ resource }: Props) => {
+  const name = agentName(resource);
```
lets deprecate the use of `agent` to mean resources? because people use the term resources anyways (it's more accurate) and this file is called resource.

referring the resource type `db` for example as agent is incorrect as well (an example of an agent is type [db_service](https://github.com/gravitational/teleport/blob/master/api/constants/constants.go#L112))  

we also renamed `agent` to `service` so you'll see those two to mean the same thing


history: we used the term `agent` to help distinguish between (nodes, apps, kube...etc) from [these resources](https://github.com/gravitational/teleport/blob/master/web/packages/teleport/src/services/resources/resource.ts), you'll see in the code base we have lots of things called resources b/c everything is a resource, `db_service` is a resource

i know we have types called `Agent`, perhaps we can rename those to `UnifiedResource` now (and unified resource will refer to nodes, apps, kube ...etc)? 

idk... naming is so hard ;;

bl-nero replied:
@kimlisa , thanks a lot for your explanations, this really clears the fog around these concepts. I will rename the references to agent concept in my code.

bl-nero replied:
Side note: I also really don't like the `AgentKind` name (That's not a kind, that's an agent! Or, actually, a resource). So renaming it to `UnifiedResource` indeed sounds like a good idea.

avatus committed:
Move Get to collector

kimlisa commented on web/packages/teleport/src/Resources/ResourceCard.tsx:
perhaps rename the root dir as `UnifiedResources`? b/c resources can also refer to... a role or a user

bl-nero commented on web/packages/teleport/src/Resources/ResourceCard.tsx:
```diff
@@ -0,0 +1,164 @@
+            </SingleLineBox>
+          )}
+        </Flex>
+        <div>
+          {resource.labels.map(({ name, value }) => (
+            <Label kind="secondary" mr={1}>{`${name}: ${value}`}</Label>
+          ))}
+        </div>
```
I actually didn't test it, let me check.

bl-nero replied:
(Oh, and yes, I plan to support many labels, but I need to address it separately.)

bl-nero committed:
Review comments

bl-nero committed:
Rename AgentKind and fix lint issues

avatus committed:
make unified resource cache public

bl-nero commented on web/packages/design/src/theme/sharedStyles.ts:
```diff
@@ -21,4 +21,9 @@ export const sharedStyles = {
     '0px 5px 5px -3px rgba(0, 0, 0, 0.2), 0px 8px 10px 1px rgba(0, 0, 0, 0.14), 0px 3px 14px 2px rgba(0, 0, 0, 0.12)',
     '0px 3px 5px -1px rgba(0, 0, 0, 0.2), 0px 6px 10px rgba(0, 0, 0, 0.14), 0px 1px 18px rgba(0, 0, 0, 0.12)',
   ],
+  breakpoints: {
```
Resolved, since Lisa accidentally ran a different broken build. :)

avatus committed:
Move to common interface

kimlisa commented on web/packages/teleport/src/UnifiedResources/ResourceCard.tsx:
```diff
@@ -0,0 +1,172 @@
+      <ResourceIcon
+        alignSelf="center"
+        name={resIcon}
+        width="60px"
+        height="60px"
```
you probably followed the figma design, but i think smaller icon works better, we can fit a few extra letters 😅 and looks less "in your face"

what do you think of `45px` instead?

bl-nero replied:
45px works fine, changed.

kimlisa commented on web/packages/teleport/src/UnifiedResources/ResourceCard.tsx:
```diff
@@ -0,0 +1,172 @@
+    case 'kube_cluster':
+      return { primary: 'Kubernetes' };
+    case 'node':
+      // TODO: Pass the subkind to display as the primary and push addr to
```
```suggestion
      // TODO(bl-nero): Pass the subkind to display as the primary and push addr to
```

kimlisa commented on web/packages/design/src/theme/sharedStyles.ts:
```diff
@@ -14,11 +14,18 @@ See the License for the specific language governing permissions and
     '0px 3px 5px -1px rgba(0, 0, 0, 0.2), 0px 6px 10px rgba(0, 0, 0, 0.14), 0px 1px 18px rgba(0, 0, 0, 0.12)',
   ],
+  breakpoints: {
+    mobile: 400 + sidebarWidth,
+    tablet: 800 + sidebarWidth,
+    desktop: 1200 + sidebarWidth,
```
ohhhh i think i get it now... so i'm not sure we should do it this way... if you look at [MainContainer.tsx](https://github.com/gravitational/teleport/blob/90774ac86dd2106edf46c8ef3d164a10fd7563f4/web/packages/teleport/src/Main/MainContainer.tsx#L28), that's where we define the side bar width, for all we know this value can change later, so i think we should be more intentional `props.theme.breakpoints.tablet + var(--sidebar-width)` (i think it should work, have not used `var()` myself yet)

however, our app atm is not designed to support smaller screens (doesn't mean we let things break like... text smashing together so it's unreadable). our main content space is set to a [min-width](https://github.com/gravitational/teleport/blob/90774ac86dd2106edf46c8ef3d164a10fd7563f4/web/packages/teleport/src/Main/Main.tsx#L261), so i don't think your breakpoints will ever apply...

bl-nero replied:
I have another branch with actual resource list screen, and it somehow _does_ apply. Can we skip it for now and revisit this thread once this part gets through and we get to the resource list?

avatus committed:
Use copy as value

avatus committed:
Move saml getter

bl-nero commented on web/packages/teleport/src/UnifiedResources/ResourceCard.tsx:
```diff
@@ -0,0 +1,172 @@
+  return (
+    <CardContainer p={3} alignItems="start">
+      <CheckboxInput type="checkbox" mx={0}></CheckboxInput>
+      <ResourceIcon
```
I don't know, perhaps it has something to do with aspect ratios. I think this sort of detail can be fixed later, I don't want to spend too much time on dealing with it right now.

roraback replied:
This is likely my fault—I think I undersized the generic icons a bit, which is throwing off the scale. I'd hoped we could lean on 3rd party icons more, and use the generic as a last resort for self-hosted resources where we can't even identify a cloud provider... As such, I wasn't using them in my designs, and didn't catch that they were smaller than the others.

bl-nero commented on web/packages/teleport/src/services/agents/types.ts:
I'll deal with it later, the project is on a tight deadline.

bl-nero committed:
Review comments

bl-nero committed:
todo

avatus committed:
invert watcher and cache

avatus committed:
use cache instead of watcher

avatus committed:
Fix imports

avatus committed:
Fix lint: check for err

avatus committed:
review round 1

avatus committed:
Update test

bl-nero committed:
Merge branch 'michaelmyers/unified_grpc' into bl-nero/unified-resources-card

avatus committed:
Add copy to get

avatus committed:
Add CloneAny interface to help with unified resource cloning

avatus committed:
Use cloneany

avatus committed:
Merge branch 'michaelmyers/unified_resource_watcher' into michaelmyers/unified_grpc

avatus committed:
Add sort for name and kind

avatus committed:
Add app and saml, fix tests

avatus committed:
Watch kube server instead of cluster

avatus committed:
Watch kube server instead of cluster

avatus committed:
Merge branch 'master' into michaelmyers/unified_resource_watcher

avatus committed:
Get kubecluster from server

avatus committed:
merge master

avatus committed:
update proto file

avatus committed:
Add no access test case

avatus committed:
move sorting

avatus committed:
Fix sorting by name

bl-nero committed:
Merge remote-tracking branch 'origin/michaelmyers/unified_grpc' into bl-nero/unified-resources-card

avatus committed:
Review feedback round 2

avatus committed:
Merge branch 'michaelmyers/unified_resource_watcher' into michaelmyers/unified_grpc

avatus committed:
Remove unneeded bk return

avatus committed:
Merge branch 'michaelmyers/unified_resource_watcher' into michaelmyers/unified_grpc

avatus committed:
Keep grpc naming convention

avatus committed:
Add missing docs

avatus committed:
defer log

bl-nero committed:
Merge branch 'michaelmyers/unified_grpc' into bl-nero/unified-resources-card

avatus committed:
Review feedback

bl-nero committed:
Merge remote-tracking branch 'origin/michaelmyers/unified_grpc' into bl-nero/unified-resources-card

avatus committed:
Add ListUnifiedResources gRPC and web endpoints (#29661)

Adds the `ListUnifiedResources` grpc endpoint that returns paginated unified resources and exposes it in the web apiserver

bl-nero committed:
Merge remote-tracking branch 'origin/michaelmyers/unified_resource_watcher' into bl-nero/unified-resources-card

bl-nero committed:
Merge remote-tracking branch 'origin/master' into bl-nero/unified-resources-card
